staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/vault/client.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package vault implements envelop encryption provider based on Vault KMS
+package vault
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/vault/api"
+)
+
+// Handle all communication with Vault server.
+type clientWrapper struct {
+	client      *api.Client
+	encryptPath string
+	decryptPath string
+	authPath    string
+
+	// We may update token for api.Client, but there is no sync for api.Client.
+	// Read lock for encrypt/decrypt requests, write lock for login requests which
+	// will update token for api.Client.
+	rwmutex sync.RWMutex
```

Comment:
Should it support renewable tokens as "first class" though instead of defaulting to a refresh approach? That seems like the best performance especially in a large cluster and the least complexity. (no locking required). If not now, I imagine we would want to in the future, so thinking about that in the current design would help make that easier. If you moved the refresh logic into the client you could have a `refreshClientWrapper` and in the future we could add a `renewalClientWrapper` with the same interface that's constructed dynamically based on info returned on the token.